pkg/controller/shared_state.go

```diff
@@ -79,25 +84,52 @@ type SharedState struct {
 	azureDiskCSITranslator        csitranslator.InTreePlugin
 }
 
-func NewSharedState(driverName, objectNamespace, topologyKey string, eventRecorder record.EventRecorder, cachedClient client.Client, azClient azdisk.Interface, kubeClient kubernetes.Interface, crdClient crdClientset.Interface) *SharedState {
+func NewSharedState(ctx context.Context, driverName, objectNamespace, topologyKey string, eventRecorder record.EventRecorder, cachedClient client.Client, azClient azdisk.Interface, kubeClient kubernetes.Interface, crdClient crdClientset.Interface) *SharedState {
 	newSharedState := &SharedState{
-		driverName:      driverName,
-		objectNamespace: objectNamespace,
-		topologyKey:     topologyKey,
-		eventRecorder:   eventRecorder,
-		cachedClient:    cachedClient,
-		crdClient:       crdClient,
-		azClient:        azClient,
-		kubeClient:      kubeClient,
-		conditionWatcher: watcher.New(context.Background(),
-			azClient, azdiskinformers.NewSharedInformerFactory(azClient, consts.DefaultInformerResync), objectNamespace),
+		driverName:             driverName,
+		objectNamespace:        objectNamespace,
+		topologyKey:            topologyKey,
+		eventRecorder:          eventRecorder,
+		cachedClient:           cachedClient,
+		crdClient:              crdClient,
+		azClient:               azClient,
+		kubeClient:             kubeClient,
+		conditionWatcher:       watcher.New(ctx, azClient, azdiskinformers.NewSharedInformerFactory(azClient, consts.DefaultInformerResync), objectNamespace),
 		azureDiskCSITranslator: csitranslator.NewAzureDiskCSITranslator(),
 	}
+
+	if crdClient != nil {
```

Comment:
We can rely on `WatchObject` function to do this instead of creating a new shared informer factory for CRD object; it's a recent addition from this commit (https://github.com/kubernetes-sigs/azuredisk-csi-driver/commit/66e996a2b40de6ffdd250a3492f2d9dbf79a2c69#diff-9af1228cdbda651b5dd579508719c1815e2297e0a4ed5326d82e196eb7eb6972). Using this, we can simplify things by setting up a watch with controller runtime package and rely on cached reader to fetch object states as well. You probably will need to modify the function to allow for predicate parameters so that we can add appropriate eventhandlers for each event.

https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/6bb1a49e3c3351856c6d572da92801472ce8b836/pkg/controller/common.go#L1105.